exchanges/gateio/gateio.go

```diff
@@ -421,169 +933,2821 @@ func (g *Gateio) SendAuthenticatedHTTPRequest(ctx context.Context, ep exchange.U
+		if err := common.StartEndTimeCheck(from, to); err != nil {
+			params.Set("to", strconv.FormatInt(to.Unix(), 10))
+		}
+	}
+	var depositHistories []DepositRecord
+	return depositHistories, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot,
+		http.MethodGet, walletDeposits, params, nil, &depositHistories)
+}
+
+// TransferCurrency Transfer between different accounts. Currently support transfers between the following:
+// spot - margin, spot - futures(perpetual), spot - delivery
+// spot - cross margin, spot - options
+func (g *Gateio) TransferCurrency(ctx context.Context, arg *TransferCurrencyParam) (*TransactionIDResponse, error) {
+	if arg == nil {
+		return nil, errNilArgument
+	}
+	if arg.Currency.IsEmpty() {
+		return nil, currency.ErrCurrencyCodeEmpty
+	}
+	if !strings.EqualFold(arg.From, asset.Spot.String()) {
+		return nil, fmt.Errorf("%w, only %s accounts can be used to transfer from", errInvalidAssetType, asset.Spot)
+	}
+	if !g.isAccountAccepted(arg.To) {
+		return nil, fmt.Errorf("%w, only %v,%v,%v,%v,%v,and %v are supported", errInvalidAssetType, asset.Spot, asset.Margin, asset.Futures, asset.DeliveryFutures, asset.CrossMargin, asset.Options)
+	}
+	if arg.Amount < 0 {
+		return nil, errInvalidAmount
+	}
+	var response *TransactionIDResponse
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodPost, walletTransfer, nil, &arg, &response)
+}
+
+func (g *Gateio) isAccountAccepted(account string) bool {
+	if account == "" {
+		return false
+	}
+	acc, err := asset.New(account)
+	if err != nil {
+		return false
+	}
+	return acc == asset.Spot || acc == asset.Margin || acc == asset.CrossMargin || acc == asset.Futures || acc == asset.DeliveryFutures || acc == asset.Options
+}
+
+func (g *Gateio) assetTypeToString(acc asset.Item) string {
+	if acc == asset.Options {
+		return "options"
+	}
+	return acc.String()
+}
+
+// SubAccountTransfer to transfer between main and sub accounts
+// Support transferring with sub user's spot or futures account. Note that only main user's spot account is used no matter which sub user's account is operated.
+func (g *Gateio) SubAccountTransfer(ctx context.Context, arg SubAccountTransferParam) error {
+	if arg.Currency.IsEmpty() {
+		return currency.ErrCurrencyCodeEmpty
+	}
+	if arg.SubAccount == "" {
+		return errInvalidOrEmptySubaccount
+	}
+	arg.Direction = strings.ToLower(arg.Direction)
+	if arg.Direction != "to" && arg.Direction != "from" {
+		return errInvalidTransferDirection
+	}
+	if arg.Amount <= 0 {
+		return errInvalidAmount
+	}
+	if arg.SubAccountType != "" && arg.SubAccountType != asset.Spot.String() && arg.SubAccountType != asset.Futures.String() && arg.SubAccountType != asset.CrossMargin.String() {
+		return fmt.Errorf("%v; only %v,%v, and %v are allowed", errInvalidAssetType, asset.Spot, asset.Futures, asset.CrossMargin)
+	}
+	return g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodPost, walletSubAccountTransfer, nil, &arg, nil)
```

Comment:
If that's the case, should I include status code 203 too?
Because it's also a **success** message even if the enclosed payload has been modified by a transforming proxy from that of the origin server's 200 ( OK ) response.


So, the update will be like this

`if resp.StatusCode < http.StatusOK ||
			resp.StatusCode > http.StatusNoContent {`

Reply:
Sounds good! Thank you!